### PR TITLE
Replace fixedwing ECL rate controller with RateControl Library

### PR DESCRIPTION
**Pull request by Jaeyoung-Lim**
## Describe problem solved by this pull request
Yaw rate was previously not controlled for fixed wings. This made it hard to integrate offboard control methods that might want to control body rates (and eventually yaw rates) directly.

This is due to the fact that fixed wing attitude control uses separate ECL controllers for each axis (roll, pitch, yaw) and specifically yaw rate control is only calculated using the current roll to maintain coordinated flight.

## Describe your solution
This commit makes the `fw_att_control` share the rate controller as a library with the `mc_rate_control` module.

- Enables passing yaw rate setpoints directly in offboard mode
- Allows controlling yaw rate setpoints in acro mode through the sticks
- Allows controlling yaw rate setpoints in offboard mode
- Removes euler body rate control from the ECL controllers
- By using the rate control library, the fw attitude controller now uses angular acceleration estimates for rate control

The attitude control and turn coordination is still done through ECL controllers

- [x]  Add airspeed scaling
- [x]  Improve tracking performance (Done by @sfuhrer )

## Describe possible alternatives
A clear and concise description of alternative solutions or features you've considered.

## Test data / coverage
- Tested in SITL Gazebo: [flight log](https://review.px4.io/plot_app?log=9a7a4bf1-fe63-4766-b1c5-250086de104a)
```
make px4_sitl gazebo_plane
```
- NOT tested on hardware

## Additional context

fixes https://github.com/PX4/PX4-Autopilot/issues/12886

### CODE REVIEW

**Comment by Jaeyoung-Lim** [comment]
I don't think the CI failures are related to this PR:
```
[0/2] Generating full parameter metadata (markdown, xml, and json)
Traceback (most recent call last):
  File "/__w/PX4-Autopilot/PX4-Autopilot/Tools/serial/generate_config.py", line 274, in <module>
    raise Exception("Default Port {:} not found for {:}".format(default_port_str, serial_command['label']))
Exception: Default Port 0 not found for Sagetech MXS Serial Port
FAILED: CMakeFiles/metadata_parameters 
```

File: src/modules/fw_att_control/FixedwingAttitudeControl.cpp
Comment (bresch) [inline src/modules/fw_att_control/FixedwingAttitudeControl.cpp:545]: The FF should be scaled with 1/TAS and the PI with 1/IAS^2. https://docs.px4.io/v1.12/en/flight_stack/controller_diagrams.html#airspeed-scaling

**Review by bresch — CHANGES_REQUESTED** [review]

File: src/modules/fw_att_control/FixedwingAttitudeControl.cpp
Comment (Jaeyoung-Lim) [reply src/modules/fw_att_control/FixedwingAttitudeControl.cpp:545]: Does this mean that rate controller essentially needs to have airspeed included in the controller calculation and we cannot unify the rate controller with multicopters? I was hoping that the airspeed scaling can be handed off to the allocation, and the rate controller only outputs "desired moments"

Comment (tstastny) [reply src/modules/fw_att_control/FixedwingAttitudeControl.cpp:545]: airspeed scaling effects each of the gains in a slightly different way (for units), right? could we externalize the CAS vs TAS calculations outside and set the scaler on an individual basis on each signal? Other option -- have no concept of "scaler" on the signals, and just input an adapted (scaled) gain from outside. wdyt? would be good to generalize this stuff as much as possible. think e.g. if we have other scaling factors later.. do we need a separate consideration for every single one in the rate controller itself? or can we just say one of the inputs to the control method is the gains, and we handle appropriately scaling those gains for whatever specific case externally.

Comment (tstastny) [reply src/modules/fw_att_control/FixedwingAttitudeControl.cpp:545]: Actually now i understand what @Jaeyoung-Lim was talking about (sorry - read through too quick) - I agree - why can't we put airspeed scaling on the actuators themselves in allocation? looking at the moment that's commanded, and then sussing out per our flight condition how much moment we actually expect to generate with a given actuator.

Comment (sfuhrer) [reply src/modules/fw_att_control/FixedwingAttitudeControl.cpp:558]: Honestly I would start by finding out how much of an effect the different scaling on FF has - would volunteer to fly a couple of planes with no distinction between FF and P/I scaling. 
And if we need to keep it then the allocation seems to be the right place for it, yes.

File: src/modules/fw_att_control/fw_att_control_params.c
Comment (sfuhrer) [inline src/modules/fw_att_control/fw_att_control_params.c:102]: Not sure if this explanation is helpful, I would rather go with something like for the MC D terms, so:

Pitch rate differential gain. Small values help reduce fast oscillations. If value is too big oscillations will appear again.

For sure try to avoid naming specific actuators here, as there doesn't need to be an elevator controlling the pitch of a vehicle.

File: src/modules/fw_att_control/fw_att_control_params.c
Comment (sfuhrer) [inline src/modules/fw_att_control/fw_att_control_params.c:107]: 1 sounds pretty extreme..but we can also first get some experience.

File: src/modules/fw_att_control/fw_att_control_params.c
Comment (sfuhrer) [inline src/modules/fw_att_control/fw_att_control_params.c:112]:
```suggestion
PARAM_DEFINE_FLOAT(FW_PR_D, 0.f);
```

**Review by sfuhrer — COMMENTED** [review]
Looks good to me so far! I plan to fly it this week, let's see if I notice performance issues (e.g. due to the different FF scaling).
I think `.setSaturationStatus() `should also be added to the fw_rate controller to preserve the anti windup ability.

File: src/lib/rate_control/CMakeLists.txt
Comment (Jaeyoung-Lim) [reply src/lib/rate_control/CMakeLists.txt:35]: Just because file name of most of the shared file names were lower case. (Since I moved the file into the shared lib and outside of the mc_rate_control). The class is still camel cased as is with other libraries. I can change it if we want to preserve the previous file name

**Comment by Jaeyoung-Lim** [comment]
- With feedforward
image

File: src/modules/fw_att_control/FixedwingAttitudeControl.cpp
Comment (sfuhrer) [reply src/modules/fw_att_control/FixedwingAttitudeControl.cpp]: addressed with d3d1e43ab03a257ca2ad03a1c2c977a28813f5c7

**Comment by sfuhrer** [comment]
I've rabased an current main and then addressed a couple of issues/missing things I found while testing it in SITL and on a real plane. See commit history.

**Comment by sfuhrer** [comment]
Flight tested again. Couldn't find any issue anymore, and while the rate and attitude tracking isn't yet prefect it was as good as it ever was on this plane, and certainly not worse than with the old, separate FW rate controller. 
https://review.px4.io/plot_app?log=d5b9af63-6e7d-42ae-90ef-99bf8488c12f
image

**Comment by Jaeyoung-Lim** [comment]
@sfuhrer Thanks for all the testing! Great that you found all the edge cases :smile:

File: src/modules/fw_att_control/fw_att_control_params.c
Comment (Jaeyoung-Lim) [reply src/modules/fw_att_control/fw_att_control_params.c]: Done!

File: src/modules/fw_att_control/FixedwingAttitudeControl.cpp
Comment (bresch) [inline src/modules/fw_att_control/FixedwingAttitudeControl.cpp:461]: Those two parts are identical except the control allocator status instance, right?
```suggestion
			if (_control_allocator_status_subs[_vehicle_status.is_vtol ? 1 : 0].update(&control_allocator_status)) {
				Vector<bool, 3> saturation_positive;
				Vector<bool, 3> saturation_negative;

				if (!control_allocator_status.torque_setpoint_achieved) {
					for (size_t i = 0; i < 3; i++) {
						if (control_allocator_status.unallocated_torque[i] > FLT_EPSILON) {
							saturation_positive(i) = true;

						} else if (control_allocator_status.unallocated_torque[i] < -FLT_EPSILON) {
							saturation_negative(i) = true;
						}
					}
				}

					// TODO: send the unallocated value directly for better anti-windup
				_rate_control.setSaturationStatus(saturation_positive, saturation_negative);
			}
```

**Review by bresch — CHANGES_REQUESTED** [review]
Great work, just one minor comment on copy-paste reduction

**Comment by sfuhrer** [comment]
> Great work, just one minor comment on copy-paste reduction

You're absolutely right ofc, that was pure laziness. Committed your proposal.

**Comment by Jaeyoung-Lim** [comment]
@sfuhrer style checks failing!

File: src/lib/rate_control/CMakeLists.txt
Comment (tstastny) [reply src/lib/rate_control/CMakeLists.txt:35]: @sfuhrer @Jaeyoung-Lim - also -  https://google.github.io/styleguide/cppguide.html#File_Names, if we are going to try and stick to the new guide

**Review by tstastny — APPROVED** [review]
great work @Jaeyoung-Lim @sfuhrer! thanks for taking this on, it is already looking lots cleaner :) i'm approving, but would be nice if you could consider my couple small comments in this review before merging.

### some thoughts on next steps.. not that need to be handled in this PR:
#### airspeed scaling
we should probably have a follow up that puts these airspeed scalings in a more appropriate place. what we really should be generating from the "rate controller" is an angular acceleration setpoint. the current formulation https://github.com/PX4/PX4-Autopilot/blob/2d2eabd41c32796f781340190ebdb11da06ad3fb/src/modules/fw_att_control/FixedwingAttitudeControl.cpp#L571-L592 assumes that we are using aerodynamic surfaces to generate the torque (hence the dynamic pressure scaling). what if this particular fixed-wing platform also wants to use differential throttle with a twin engine for yaw torque (or *all torques, for that matter.. e.g. K1 Swan)? the airspeed scaling will be different, e.g. advance ratio considerations for prop inflow, linear with inflow, not quadratic.

main takeaway is that airspeed scaling is not a rate control issue, but tied to the actuator that is either already generating, or will be allocated for generating, the desired moment. so better to scale things given knowledge of the actual actuator type that will be used, than here in the rate control and disappearing into the torque setpoint. desired angular acceleration doesnt change with airspeed. required actuation to achieve it does.

#### wheel control
looks like the ECL attitude control class could be much simpler minus the wheel controller.. maybe we can break this off into it's own class.. and also give it a much needed revisit

#### rate control module
if we are using the rate control library for both multicopter and fixed-wing, could we now break it into it's own library?? then VTOL has a consistent rate controller running in **all** modes!

#### merging attitude control
next step.. use quaternion attitude control commonly for both fixed-wing and MC as well? :)

File: src/modules/fw_att_control/FixedwingAttitudeControl.cpp
Comment (sfuhrer) [reply src/modules/fw_att_control/FixedwingAttitudeControl.cpp:522]: There is no longer a separate integrator reset for the different axis. We either reset all 3 or none. And that reset happens always when `landed` or when t`akeoff_state_ < RunwayTakeoffState::CLIMBOUT`, so I think we are covered.

File: src/modules/fw_att_control/ecl_pitch_controller.h
Comment (sfuhrer) [reply src/modules/fw_att_control/ecl_pitch_controller.h:70]: done in d6e7ca98a17ad87

File: src/modules/fw_att_control/FixedwingAttitudeControl.cpp
Comment (sfuhrer) [reply src/modules/fw_att_control/FixedwingAttitudeControl.cpp]: done 6b77b59ded

**Comment by sfuhrer** [comment]
Thanks for the review and the afterthoughts @tstastny !

>main takeaway is that airspeed scaling is not a rate control issue, but tied to the actuator that is either already generating, or will be allocated for generating, the desired moment. so better to scale things given knowledge of the actual actuator type that will be used, than here in the rate control and disappearing into the torque setpoint. desired angular acceleration doesnt change with airspeed. required actuation to achieve it does.

Yes absolutely. I had the same thoughts while working on a control surface less VTOL, all the actuator specific scaling should happen in the allocator.

>looks like the ECL attitude control class could be much simpler minus the wheel controller.. maybe we can break this off into it's own class.. and also give it a much needed revisit

Indeed, could be a nice little contained follow up:)

>if we are using the rate control library for both multicopter and fixed-wing, could we now break it into it's own library?? then VTOL has a consistent rate controller running in all modes!

Single rate control module for both the MC and FW part of a VTOL flight? Absolutely on the planning sheet, but will need more work on the allocation side (except for simple tailsitter setups that use the same actuators in both modes).

>next step.. use quaternion attitude control commonly for both fixed-wing and MC as well? :)

Yes! The MC attitude controller though has some more "MC-specifics" in it than the MC rate controller, so we need to see if a single implementation makes sense. If not they should at least use a shared library.

**Comment by Jaeyoung-Lim** [comment]
@bresch @tstastny @sfuhrer Anything else to address? I think it would be great if we can get this in and try to improve the things we learned on incremental PRs!

**Review by sfuhrer — APPROVED** [review]

**Review by bresch — APPROVED** [review]

**Pull request merged**
